File: 240326/BOJ-2805.py
N, M = map(int, input().split())
tree = [int(num) for num in input().split()]

left = 0
right = max(tree)

answer = 0
candidate = 0
find = False
while left <= right:
    target = (left + right) // 2

    length = 0
    for t in tree:
        if t > target:
            length += (t - target)
    
    if length < M:
        right = target - 1
    elif length > M:
        left = target + 1
        candidate = target
    elif length == M:
        answer = target
        find = True
        break

if find == True:
    print(answer)
else:
    print(candidate)

File: 240326/BOJ-2805.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main(int argc, char** argv){
    int N, M;
    cin >> N >> M;

    vector<int> tree(N);
    for(int i=0; i<N; i++){
        cin >> tree[i];
    }

    int left = 0;
    int right = *max_element(tree.begin(), tree.end());

    int answer = 0;
    int candidate = 0;
    bool find = false;

    while(left <= right){
        int target = (left + right) / 2;

        long long length = 0;
        for(int i=0; i < N; i++){
            if(tree[i] > target){
                length += (tree[i] - target);
            }
        }

        if(length < M){
            right = target - 1;
        } else if(length > M){
            left = target + 1;
            candidate = target;
        } else if(length == M){
            answer = target;
            find = true;
            break;
        }
    }

    if(find == true){
        cout << answer;
    } else{
        cout << candidate;
    }
}